### ISSUE TITLE
Free moat (optimizations that cost ESSENTIALLY nothing)

### ISSUE THREAD

**Issue by reyvvon** [jrz97619761/test-model-thing#19]
## Training Performance Optimization

Training currently evaluates traces separately for every layer and rebuilds the AdamW update each byte. We can improve this by:

- Reusing the byte mask and decayed embedding trace.
- Evaluating all layer traces together.
- Compiling the AdamW update after its first execution.

### Benchmark

Tested on the pasted **512×16 model** using spread initialization and the scheduled learning rate, on an **Apple M5 with MLX 0.32.2**.

| Benchmark | Before | After |
|---|---:|---:|
| 5,000 forward/backward passes | 47.50 s | 19.54 s |
| 1,000-pass profile | 10.78 ms/step | 4.35 ms/step |
| Peak memory | 131 MB | 143 MB |

Ten full-size training steps matched **exactly** for:

- Outputs
- Parameters
- Optimizer state
- Recurrent traces
- Steps where `nextb=None`

Checkpoint reload behavior also matched a fresh model.

## Implementation

1. **Initialize the compiled parameter update**

   After:

```python
   self.optimizer = opt.AdamW(learning_rate=lrfn)
```

add:

```python
self._compiled_parameter_update = None
```

2. **Reuse the byte mask and decayed embedding trace**

   In the training branch of `__call__`, compute the byte mask once before the layer loop:

   ```python
   (mx.arange(256) == c)[:, None].astype(mx.float32)
   ```

   Inside the loop, save:

   ```python
   decayed_embedtrace = layer.embedtrace * decays[i]
   ```

   Reuse `decayed_embedtrace` for both `embedtrace` and the encoder gradient.

3. **Batch trace evaluation**

   Move:

   ```python
   mx.eval(layer.states, layer.decaytrace, layer.embedtrace)
   ```

   out of the layer loop.

   Evaluate `states`, `decaytrace`, and `embedtrace` for all layers in a single `mx.eval(...)` call.

4. **Compile AdamW updates after initialization**

   Move the AdamW update into:

   ```python
   Model.update_parameters(grads)
   ```

   Run the first update normally so AdamW can initialize its optimizer state.

   For subsequent updates, use `mx.compile` with:

   ```python
   inputs=[self.state, self.optimizer.state]
   ```

   and the same outputs.

   Keep:

   ```python
   mx.eval(self.parameters(), self.optimizer.state)
   ```

   after every update.

5. **Reset compiled state after loading a checkpoint**

   In `load()`, after the file-existence check, add:

   ```python
   self._compiled_parameter_update = None
   ```

   This ensures a loaded checkpoint creates a fresh compiled parameter update.

**Comment by reyvvon** [comment]
I was in a rush so I wrote the report using AI if something is not clear drop a msg!

**Comment by reyvvon** [comment]
Definitely validate the claims in your environment

**Comment by jrz97619761** [comment]
Ok, it's fine if you were in a rush @reyvvon, no worries. I do wonder where the ```decayed_embedtrace``` is used, that wasn't made clear. What's also not 100% clear for me is how I'm supposed to move the AdamW update and ```mx.compile```, I would appreciate it if that was clarified a little more. Thanks

**Comment by reyvvon** [comment]
Thanks for pointing that out. `decayed_embedtrace` is the existing `layer.embedtrace * decays[i]` calculation. It is used **twice**: to form the next trace with `+ current_byte`, and to calculate the encoder gradient with `dlds * decayed_embedtrace`. This avoids calculating the same product twice.

For AdamW, add this method inside `Model`, immediately after `step()`:

```python
def update_parameters(self, grads):
     if self._compiled_parameter_update is None:
         # The first update creates AdamW's per-parameter state.
         self.optimizer.update(self, grads)
         mx.eval(self.parameters(), self.optimizer.state)
         state = [self.state, self.optimizer.state]

         def update(gradients):
             self.optimizer.update(self, gradients)
             return self.optimizer.state["step"]

         self._compiled_parameter_update = mx.compile(
             update, inputs=state, outputs=state
         )
         return

     self._compiled_parameter_update(grads)
     mx.eval(self.parameters(), self.optimizer.state)
```

Add `self._compiled_parameter_update = None` immediately after `self.optimizer = opt.AdamW(learning_rate = lrfn)` in `__init__`. In the training part of `__call__`, replace the trace calculation and evaluation as follows:

 ```python
 current_byte = (mx.arange(256) == c)[:, None].astype(mx.float32)
 for i, layer in enumerate(self.layers):
     dlds = dlds_s[i]
     decayed_embedtrace = layer.embedtrace * decays[i]
     embedtrace = decayed_embedtrace + current_byte
     grads["encoder"]["embed"]["weight"] += dlds * decayed_embedtrace

     decaytrace = (decays[i] * layer.decaytrace) + (
         decays[i] * (1.0 - decays[i]) * layer.states
     )
     grads["layers"][i]["decay"] = dlds * decaytrace
     layer.states = mx.stop_gradient(states[i])
     layer.decaytrace = mx.stop_gradient(decaytrace)
layer.embedtrace = mx.stop_gradient(embedtrace)

mx.eval(*[
     value
     for layer in self.layers
     for value in (layer.states, layer.decaytrace, layer.embedtrace)
])
self.update_parameters(grads)
```

This replaces both the `mx.eval(...)` **inside** the loop and the two AdamW lines **after** it. Finally, in `load()`, add `self._compiled_parameter_update = None` immediately after `if not os.path.exists(path): return`, so loading a checkpoint discards the old compiled update.

**Comment by reyvvon** [comment]
Should be 2.45x faster than before per step.

**Comment by reyvvon** [comment]
@jrz97619761 any update?

**Comment by jrz97619761** [comment]
Will implement this asap

**Comment by jrz97619761** [comment]
@reyvvon implemented.

One question, can you double check if the below lines in Model.save() (and its corresponding lines in model.load() is even needed? Or does MLX save these already (my check might be wrong)? thanks. after we confirm that I'll close this issue

```python
        for i, layer in enumerate(self.blocks):
            data[f"state.{i}"] = layer.states
            data[f"decaytrace.{i}"] = layer.decaytrace
            data[f"embedtrace.{i}"] = layer.embedtrace
```
